src/meta/src/stream/scale.rs

```diff
@@ -0,0 +1,824 @@
+
+            for actor in &fragment.actors {
+                if let Some(upstream_actor_tuples) = ctx.upstream_dispatchers.get(&actor.actor_id) {
+                    for (upstream_fragment_id, _, upstream_dispatcher_id) in upstream_actor_tuples {
+                        upstream_fragment_dispatcher_set
+                            .insert((*upstream_fragment_id, *upstream_dispatcher_id));
+                    }
+                }
+            }
+
+            let downstream_fragment_id = if let Some(downstream_fragment_ids) =
+                ctx.downstream_fragment_id_map.get(&fragment_id)
+            {
+                let downstream_fragment_id = downstream_fragment_ids.iter().exactly_one().unwrap();
+                Some(*downstream_fragment_id as FragmentId)
+            } else {
+                None
+            };
+
+            let upstream_fragment_dispatcher_ids =
+                upstream_fragment_dispatcher_set.into_iter().collect_vec();
+            reschedule_fragment.insert(
+                fragment_id,
+                Reschedule {
+                    added_actors: actors_to_create,
+                    removed_actors: actors_to_remove,
+                    added_parallel_units,
+                    removed_parallel_units,
+                    vnode_bitmap_updates: Default::default(),
+                    upstream_fragment_dispatcher_ids,
+                    upstream_dispatcher_mapping,
+                    downstream_fragment_id,
+                },
+            );
+        }
+
+        let mut fragment_created_actors = HashMap::new();
+        for (fragment_id, actors_to_create) in &fragment_actors_to_create {
+            let mut created_actors = HashMap::new();
+            for (actor_id, parallel_unit_id) in actors_to_create {
+                let actor = new_created_actors.get(actor_id).cloned().unwrap();
+                let worker_id = ctx
+                    .parallel_unit_id_to_worker_id
+                    .get(parallel_unit_id)
+                    .unwrap();
+                created_actors.insert(
+                    *actor_id,
+                    (
+                        actor,
+                        ActorStatus {
+                            parallel_unit: Some(ParallelUnit {
+                                id: *parallel_unit_id,
+                                worker_node_id: *worker_id,
+                            }),
+                            state: ActorState::Inactive as i32,
+                        },
+                    ),
+                );
+            }
+
+            fragment_created_actors.insert(*fragment_id, created_actors);
+        }
+
+        self.fragment_manager
+            .pre_apply_reschedules(fragment_created_actors)
+            .await?;
+
+        self.barrier_manager
+            .run_multiple_commands(vec![
+                Command::Plain(Some(Mutation::Pause(PauseMutation {}))),
+                Command::RescheduleFragment(reschedule_fragment),
```

Comment:
What if `RescheduleFragment` fails, will the cluster enter normal status? 🤔 I guess not, if `RescheduleFragment` fails, a recovery process will triggered and command with `ResumeMutation` will be aborted. But in this situation, all `SourceExecutor`s are paused, the recovery process will hang. This's not related to this PR, I think we should take `Pause` status into consideration for recovery. Cc @BugenZhao

Reply:
> a recovery process will triggered and command with ResumeMutation will be aborted

If the recovery is triggered, will the `Resume` command be removed? If so and we rebuild the actors, I think the source executor will be resumed initially. 🤣

However, this comment makes sense since this interface only ensures the issuing of these commands are atomic, not the executing. I think we may improve this later.

Reply:
Yes, the `Resume` command will be removed. The problem is when in recovery, the stop barrier can't flow through the source executors, because they're paused.

Reply:
The `pause` barrier only pauses the chunk side and leave the barrier side untouched. So I think there's no much problem?

Reply:
Oh yes, the barrier still can flow. Well, that will be no problem.